src/client.rs

```diff
@@ -5,20 +5,27 @@ use proto::{self, Connection, WindowSize};
 use error::Reason::*;
 
 use http::{Request, Response};
-use futures::{Future, Poll, Sink, Async, AsyncSink};
+use futures::{Future, Poll, Sink, Async, AsyncSink, AndThen, MapErr};
 use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_io::io::WriteAll;
 use bytes::{Bytes, IntoBuf};
 
 use std::fmt;
+use std::io::Error as IoError;
 
 /// In progress H2 connection binding
-pub struct Handshake<T, B: IntoBuf = Bytes> {
-    // TODO: unbox
-    inner: Box<Future<Item = Client<T, B>, Error = ConnectionError>>,
+pub struct Handshake<T: AsyncRead + AsyncWrite, B: IntoBuf = Bytes> {
+    inner:
+        AndThen<
+            MapErr<WriteAll<T, &'static [u8]>, fn(IoError) -> ConnectionError>,
+            Result<Client<T, B>, ConnectionError>,
+            fn((T, &'static [u8])) -> Result<Client<T, B>, ConnectionError>
+        >
+
 }
 
 /// Marker type indicating a client peer
-pub struct Client<T, B: IntoBuf> {
+pub struct Client<T: AsyncRead + AsyncWrite, B: IntoBuf> {
```

Comment:
Adding this bound is fine w/ me. I'm just curious if this became required due to the change or it was added for some other reason.

Reply:
If I recall correctly, this bound was necessary for this change in 658c80b6a9535b4a5645b4c29f5275ca516cec28. I'm not sure if it's still necessary as of a67fd958fdc1977c1d4f8438dd2e898cf78680c5?